hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpoint.java

```diff
@@ -17,40 +17,119 @@
 package com.google.cloud.bigtable.hbase2_x.replication;
 // TODO reverse the package names for 1.x and 2.x to com.google.cloud.bigtable.replication.hbase1_x
 
+import static com.google.cloud.bigtable.hbase2_x.replication.configuration.CustomWALEntryConfig.REPLICATION_WALENTRYFILTER_CONFIG_KEY;
+
 import com.google.bigtable.repackaged.com.google.api.core.InternalExtensionOnly;
 import com.google.cloud.bigtable.hbase.replication.CloudBigtableReplicator;
 import com.google.cloud.bigtable.hbase.replication.adapters.BigtableWALEntry;
 import com.google.cloud.bigtable.hbase2_x.replication.metrics.HBaseMetricsExporter;
+import java.io.IOException;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.UUID;
-import org.apache.hadoop.hbase.replication.BaseReplicationEndpoint;
+import org.apache.hadoop.hbase.replication.ChainWALEntryFilter;
+import org.apache.hadoop.hbase.replication.NamespaceTableCfWALEntryFilter;
+import org.apache.hadoop.hbase.replication.ReplicationEndpoint;
+import org.apache.hadoop.hbase.replication.ReplicationPeer;
+import org.apache.hadoop.hbase.replication.ReplicationPeerConfig;
+import org.apache.hadoop.hbase.replication.ScopeWALEntryFilter;
+import org.apache.hadoop.hbase.replication.WALEntryFilter;
 import org.apache.hadoop.hbase.wal.WAL;
+import org.apache.hbase.thirdparty.com.google.common.collect.Lists;
+import org.apache.hbase.thirdparty.com.google.common.util.concurrent.AbstractService;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-// TODO(remove BaseReplicationEndpoint extension).
 @InternalExtensionOnly
-public class HbaseToCloudBigtableReplicationEndpoint extends BaseReplicationEndpoint {
+public class HbaseToCloudBigtableReplicationEndpoint extends AbstractService
+    implements ReplicationEndpoint {
 
   private static final Logger LOG =
       LoggerFactory.getLogger(HbaseToCloudBigtableReplicationEndpoint.class);
 
   private final CloudBigtableReplicator cloudBigtableReplicator;
   private final HBaseMetricsExporter metricsExporter;
 
+  protected Context ctx;
+
   public HbaseToCloudBigtableReplicationEndpoint() {
     cloudBigtableReplicator = new CloudBigtableReplicator();
     metricsExporter = new HBaseMetricsExporter();
   }
 
+  @Override
+  public void init(Context context) throws IOException {
+    this.ctx = context;
+    if (this.ctx != null) {
+      ReplicationPeer peer = this.ctx.getReplicationPeer();
+      if (peer != null) {
+        peer.registerPeerConfigListener(this);
+      } else {
+        LOG.warn(
+            "Not tracking replication peer config changes for Peer Id "
```

Comment:
When will this happen and what are the implications of it?

Reply:
The user of this class(ReplicationSource) must init it before using. https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSource.java#L321


It might happen if we instantiate the class and don't call it.

Reply:
please add comments explaining this reason